specs/derivation.md

```diff
@@ -27,181 +43,606 @@ TODO
+transaction_data = version_byte ++ rollup_payload
+
+# version_byte == 0:
+rollup_payload = frame ...   #  one or more frames, concatenated
 
-channel_id        = bytes32 | uvarint   # identifies the channel to add frame_data to, as well as the time (UNIX-epoch in seconds) the channel was created.
-frame_number      = uvarint  # identifies the index of frame_data in the channel
-frame_data_length = uvarint  # frame_data_length == len(frame_data)
-frame_data        = bytes    # data to add to the channel, after frame max(frame_number-1, 0)
-is_last           = bool     # 1 byte, channel is closed if 1, stays open if 0, invalid otherwise
+# version_byte != 0: invalid, reserved for future use
 ```
 
-The channel identity uniqueness is limited both by the random `bytes32` in the ID, as well as the timestamp.
-The timestamp is limited by L1 timestamps, to force-close a channel and avoid revival of the same channel ID after the `CHANNEL_TIMEOUT`.
-This ensures at most a span of `CHANNEL_TIMEOUT` seconds of L1 data is required to reconstruct any channel-bank state in time.
+(where `++` denotes concatenation)
 
-### L1 transaction data format:
+The `rollup_payload` may be right-padded with 0s, which will be ignored. It's allowed for them to be
+interpreted as frames for channel 0, which must always be ignored.
+
+> **TODO** specify batcher authentication (i.e. where do we store / make available the public keys of authorize batcher
+> signers)
+
+### Frame Format
+
+A [channel frame][g-channel-frame] is encoded as:
 
 ```text
-transaction_data = version_byte | rollup_payload
+frame = channel_id ++ frame_number ++ frame_data_length ++ frame_data ++ is_last
+
+channel_id        = random ++ timestamp
+random            = bytes32
+timestamp         = uvarint
+frame_number      = uvarint
+frame_data_length = uvarint
+frame_data        = bytes
+is_last           = bool
+```
 
-# version_byte == 0:
-rollup_payload = frame ...   #  one or more frames, concatenated
+where:
+
+- `uvarint` is a variable-length encoding of a 64-bit unsigned integer into between 1 and 9 bytes, [as specified in
+  SQLite 4][sqlite-uvarint].
+- `channel_id` uniquely identifies a channel as the concatenation of a random value and a timestamp
+    - `random` is a random value such that two channels with different batches should have a different random value
+    - `timestamp` is the time at which the channel was created (UNIX time in seconds)
+    - The ID includes both the hash and the timestamp, in order to prevent a malicious sequencer from reusing the random
+      value after the channel has [timed out][g-channel-timeout] (refer to the [batcher specification][batcher-spec] to
+      learn more about channel timeouts). This will also allow us substitute `random` by a hash commitment to the
+      batches, should we want to do so in the future.
+- `frame_number` identifies the index of the frame within the channel (in bytes)
```

Comment:
incorrect, it's an index not an offset
similarly `max(frame_number-1, 0)` is incorrect — should talk about it being after he previous frames